### ISSUE TITLE
Inpainting vs text conditioned inpainting vs Paint-by-example

### ISSUE THREAD

**Issue by ivineetm007** [buxiangzhiren/Asymmetric_VQGAN#4]
Hi,
Thanks for sharing your work on Git Hub. I am a bit confused with the different models provided and how to run them?
**Inpainting**- Given an image and mask, fill the masked area. Similar to the inpainting model in compvis/latent-diffusion repo and Lama as compared in your paper. In the latent diffusion repo, the model is provided here https://github.com/CompVis/latent-diffusion#inpainting. It is different from sdv1.5 inpainting.

**Text-guided inpainting**- Given an image, mask, and text. Fill the masked area based on the text guidance. Similar to sd-v1-5-inpainting shared in the runwayml repo https://github.com/runwayml/stable-diffusion

**Paint-by-example**- Given a reference image, source image and the corresponding mask. Fill the masked area based on the reference image.

The inference script 'inpaint_st.py'. It is loading image and mask from val.rar. Also, using hard-coded text in the script. 
For which problem setting, this script is written? How to run the inference for all the three settings? What are the models required to run inference of all these problem settings?

**Comment by buxiangzhiren** [comment]
The inference script 'inpaint_st.py is written in the text-guided inpainting task. As to the inpainting task, we follow the setting in the runwayml repo https://github.com/runwayml/stable-diffusion which use a fixed prompt of _photograph of a beautiful empty scene, highest quality settings_

**Comment by buxiangzhiren** [comment]
> The inference script 'inpaint_st.py is written in the text-guided inpainting task. As to the inpainting task, we follow the setting in the runwayml repo https://github.com/runwayml/stable-diffusion which use a fixed prompt of _photograph of a beautiful empty scene, highest quality settings_

So sd-v1-5-inpainting model is used for both the inpainting and the text-guided inpainting tasks.

**Comment by buxiangzhiren** [comment]
> The inference script 'inpaint_st.py is written in the text-guided inpainting task. As to the inpainting task, we follow the setting in the runwayml repo https://github.com/runwayml/stable-diffusion which use a fixed prompt of _photograph of a beautiful empty scene, highest quality settings_

The fixed prompt is here https://github.com/buxiangzhiren/Asymmetric_VQGAN/blob/main/inpaint_st.py#LL185C9-L185C83

**Comment by buxiangzhiren** [comment]
Last, this script does not support the paint-by-example task. In fact, we use the model (https://drive.google.com/file/d/15QzaTWsvZonJcXsNv-ilMRCYaQLhzR_i/view?pli=1) provided by the official repository (https://github.com/Fantasy-Studio/Paint-by-Example).

**Comment by vinm007** [comment]
The inpainting model is different from text-guided inpainting. For inpainting, this model was used https://heibox.uni-heidelberg.de/f/4d9ac7ea40c64582b7c9/?dl=1 as mentioned in latent-diffusion repo. For inpainting inference, they mentioned `inpaint.py` https://github.com/CompVis/latent-diffusion/blob/main/scripts/inpaint.py for inference which doesn't use  fixed prompt as model doesn't take text as input.
In short, there are two scripts they used 'inpaint.py' for inpainting with this model and 'inpaint_st.py' for text guided inpainting for which sdv1-5-inpainting is used.
The latent diffusion paper compare results for inpainting only probably using this model https://heibox.uni-heidelberg.de/f/4d9ac7ea40c64582b7c9/?dl=1 
How did you compared your results for inpainting?

**Comment by buxiangzhiren** [comment]
> Last, this script does not support the paint-by-example task. In fact, we use the model (https://drive.google.com/file/d/15QzaTWsvZonJcXsNv-ilMRCYaQLhzR_i/view?pli=1) provided by the official repository (https://github.com/Fantasy-Studio/Paint-by-Example).

**Comment by buxiangzhiren** [comment]
> 

We follow the setting of sdv1-5-inpainting in https://github.com/runwayml/stable-diffusion to compare our results for inpainting
![image](https://github.com/buxiangzhiren/Asymmetric_VQGAN/assets/67734862/77d90549-ca45-4cfd-9d42-8f81a46d2f7e)

**Comment by buxiangzhiren** [comment]
> > 
> 
> We follow the setting of sdv1-5-inpainting in https://github.com/runwayml/stable-diffusion to compare our results for inpainting ![image](https://user-images.githubusercontent.com/67734862/246101463-77d90549-ca45-4cfd-9d42-8f81a46d2f7e.png)

My understanding is that a fixed text input is equivalent to not needing to manually input text anymore, making text-guided inpainting consistent with traditional inpainting. Additionally, you can also use an empty string as a text input, although the inpainting result may not be as good as what the official method provides.

**Comment by vinm007** [comment]
I think it is not fairly compared.
The stable diffusion inpainting (sd-v1.5-inpainting) is different from latent diffusion inpainting. This is what runawayml has tried to compare a without text input inpainting model with their sd-v1.5-inpainting. The original paper https://arxiv.org/abs/2112.10752 probably doesn't use this strategy to compare results with Lama due to simple fact that latent diffusion repo was released along with the original paper. At that time, there was no stable diffusion.
The stable diffusion version models are not reported in the paper. Look at the paper and supplementary materials no mention of text-guided inpainting.

**Comment by buxiangzhiren** [comment]
You mentioned that runawayml has tried to compare a without text input inpainting model with their sd-v1.5-inpainting.  And our goal is to compare the performance of our asymmetric VQGAN with the original VQGAN on the stable diffusion model. We adopt the methods of the stable diffusion model to evaluate. Is there any unfair setting here?

**Comment by buxiangzhiren** [comment]
You also mentioned that the stable diffusion inpainting (sd-v1.5-inpainting) is different from latent diffusion inpainting. Our asymmetric vqgan design is specifically targeted at stable diffusion models. However, why are we evaluating it according to the methodology of latent diffusion models?

**Comment by vinm007** [comment]
Help me understand this.
Lama and latent diffusion models are both trained and evaluated on Places dataset. Reference 
![image](https://github.com/buxiangzhiren/Asymmetric_VQGAN/assets/136594182/29518558-66af-4007-8efa-2022e5b5d575)

The runawayml model is trained on different subsets of LAION dataset
![image](https://github.com/buxiangzhiren/Asymmetric_VQGAN/assets/136594182/cb083e5b-960d-4280-ab97-85cc9ee80b7e)

The runawayml model is compared along with Lama and Latent Diffusion model on Places test set. But it's not a fair evaluation (in research setting) in itself since the training set is different. I think it doesn't matter because runaway is not proposing new novel technique and wants to check if the technique led to good results or data. In a real world application it doesn't matter what the training set is.

But for the research setting, the training and test set should be same. Otherwise you can't be sure whether it's the model or data. Since, you proposing new technique it should be evaluated on same training set (or subset of training set.) 
Do you thing it's fair to compare Lama technique which is just trained Places to be evaluated with stable diffusion model which is trained on different dataset?

**Comment by buxiangzhiren** [comment]
Our work involves designing an asymmetric VQGAN responsible for the decoding part. The effectiveness of inpainting directly relies on the diffusion process of stable diffusion models, and we have not made any parameter modifications to the diffusion process. It can be said that the unfair comparison arises from the differences in training datasets caused by the use of different diffusion processes in stable diffusion. However, this is not something we can decide. Furthermore, the VQGAN in latent diffusion is also trained on the ImageNet dataset. Following your argument, would it be unfair to compare latent diffusion with LAMA as well?

**Comment by vinm007** [comment]
> Our work involves designing an asymmetric VQGAN responsible for the decoding part. The effectiveness of inpainting directly relies on the diffusion process of stable diffusion models, and we have not made any parameter modifications to the diffusion process. It can be said that the unfair comparison arises from the differences in training datasets caused by the use of different diffusion processes in stable diffusion. However, this is not something we can decide. Furthermore, the VQGAN in latent diffusion is also trained on the ImageNet dataset. Following your argument, would it be unfair to compare latent diffusion with LAMA as well?

In latent diffusion, they have different models trained on different datasets for different tasks. Can you share any reference where it says, for inpainting training is first done on ImageNet and then on Places.
As per the supplementary material, ImageNet is used for Class-Label to Image and Superresolution
![image](https://github.com/buxiangzhiren/Asymmetric_VQGAN/assets/136594182/7add5a8c-c5f2-4f46-8843-d4fb25e27099)

**Comment by vinm007** [comment]
I am not sure whether the default setting indicates the actual thing.
Also, why did they later train and tested on Places only? They could have used any large dataset for training and compared it against the Places. In fact, they are the ones who led to the creation of LAION datasets. They could have used that and got very impressive results in terms of numbers. But still, as per the paper, they trained (let's say second stage only) and tested on Places for a fair evaluation. At least some attempt is there for a fair evaluation.

**Comment by buxiangzhiren** [comment]
We also want to design a fair comparison. Do you think it's fair to evaluate our asymmetric VQGAN against the original VQGAN as the evaluation of stable diffusion?

**Comment by buxiangzhiren** [comment]
> If we change our evaluation method for inpainting, which is inconsistent with the evaluation method used for stable diffusion itself, can we still fairly compare our asymmetric VQGAN?

To ensure a fair comparison of our asymmetric VQGAN, we have no choice but to follow the evaluation methodology of stable diffusion.

**Comment by buxiangzhiren** [comment]
I did not confuse the results of latent diffusion and stable diffusion. The comparison with LAMA and latent diffusion is made by stable diffusion itself, and it is unrelated to our evaluation. Our goal is to validate that our method is better than stable diffusion because of our asymmetric VQGAN. In the comparison with stable diffusion, our approach is absolutely fair.

**Comment by buxiangzhiren** [comment]
I said all this just hoping that you can understand the purpose of the experimental evaluation, which is to verify the effectiveness of asymmetric VQGAN, rather than to verify which one is better between Stable Diffusion, LAMA, and Latent Diffusion.

The fact that our performance is better than LAMA and Latent Diffusion does not necessarily demonstrate the advantage of our asymmetric VQGAN. It is because you cannot determine whether the advantage comes from Stable Diffusion or our asymmetric VQGAN.

Lastly, wich you can have a profound understanding of the purpose of the evaluation and make further advancements in your research.
